File: mobile/lib/models/loan.dart
import 'user.dart';
import 'payment.dart';

class Loan {
  final int id;
  final int lenderId;
  final int borrowerId;
  final double amount;
  final String currency;
  final String? description;
  final DateTime? dueDate;
  final String status;
  final double remainingAmount;
  final DateTime createdAt;
  final User? lender;
  final User? borrower;
  final List<Payment>? payments;

  Loan({
    required this.id,
    required this.lenderId,
    required this.borrowerId,
    required this.amount,
    required this.currency,
    this.description,
    this.dueDate,
    required this.status,
    required this.remainingAmount,
    required this.createdAt,
    this.lender,
    this.borrower,
    this.payments,
  });

  factory Loan.fromJson(Map<String, dynamic> json) {
    return Loan(
      id: json['id'],
      lenderId: json['lender_id'],
      borrowerId: json['borrower_id'],
      amount: double.parse(json['amount'].toString()),
      currency: json['currency'] ?? 'MMK',
      description: json['description'],
      dueDate: json['due_date'] != null
          ? DateTime.parse(json['due_date'])
          : null,
      status: json['status'],
      remainingAmount: double.parse(json['remaining_amount'].toString()),
      createdAt: DateTime.parse(json['created_at']),
      lender: json['lender'] != null ? User.fromJson(json['lender']) : null,
      borrower: json['borrower'] != null ? User.fromJson(json['borrower']) : null,
      payments: json['payments'] != null
          ? (json['payments'] as List).map((p) => Payment.fromJson(p)).toList()
          : null,
    );
  }

  bool get isPaid => status == 'paid';
  bool get isPending => status == 'pending';
  bool get isPartial => status == 'partial';
  bool get isOverdue => dueDate != null && dueDate!.isBefore(DateTime.now()) && !isPaid;
  bool get isDueSoon => dueDate != null &&
      dueDate!.isAfter(DateTime.now()) &&
      dueDate!.isBefore(DateTime.now().add(const Duration(days: 3))) &&
      !isPaid;
}

File: backend/app/Models/Loan.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Loan extends Model
{
    protected $fillable = [
        'lender_id',
        'borrower_id',
        'amount',
        'currency',
        'description',
        'due_date',
        'status',
        'remaining_amount',
    ];

    protected function casts(): array
    {
        return [
            'amount' => 'decimal:2',
            'remaining_amount' => 'decimal:2',
            'due_date' => 'date',
        ];
    }

    public function lender()
    {
        return $this->belongsTo(User::class, 'lender_id');
    }

    public function borrower()
    {
        return $this->belongsTo(User::class, 'borrower_id');
    }

    public function payments()
    {
        return $this->hasMany(Payment::class);
    }

    public function notifications()
    {
        return $this->hasMany(LoanNotification::class);
    }

    public function scopePending($query)
    {
        return $query->where('status', 'pending');
    }

    public function scopePaid($query)
    {
        return $query->where('status', 'paid');
    }

    public function scopeOverdue($query)
    {
        return $query->where('status', '!=', 'paid')
            ->whereNotNull('due_date')
            ->where('due_date', '<', now());
    }

    public function scopeDueSoon($query, $days = 1)
    {
        return $query->where('status', '!=', 'paid')
            ->whereNotNull('due_date')
            ->whereBetween('due_date', [now(), now()->addDays($days)]);
    }
}
